### PR TITLE
[PYTORCH][NEURON][BUILD] - PT1.8 support

### PR DESCRIPTION
**Pull request by aws-vrnatham**
Signed-off-by: Venky Natham <vrnatham@amazon.com>

*GitHub Issue #, if available:*

Note: If merging this PR should also close the associated Issue, please also add that Issue # to the Linked Issues section on the right. 

### Description
Add the torch-neuron version 1.8 of the DLC container.

### Tests run
The pytorch tests are run with this new container.

## Label Checklist
- [ ] I have added the project label for this PR (*<project_name>* or "Improvement")

## PR Checklist
- [ ] I've prepended PR tag with frameworks/job this applies to : [mxnet, tensorflow, pytorch] | [ei/neuron] | [build] | [test] | [benchmark] | [ec2, ecs, eks, sagemaker]
- [ ] If the PR changes affects SM test, I've modified dlc_developer_config.toml in my PR branch by setting sagemaker_tests = true and efa_tests = true
- [ ] If this PR changes existing code, the change fully backward compatible with pre-existing code. (Non backward-compatible changes need special approval.)
- [ ] (If applicable) I've documented below the DLC image/dockerfile this relates to
- [ ] (If applicable) I've documented below the tests I've run on the DLC image
- [ ] (If applicable) I've reviewed the licenses of updated and new binaries and their dependencies to make sure all licenses are on the Apache Software Foundation Third Party License Policy Category A or Category B license list.  See [https://www.apache.org/legal/resolved.html](https://www.apache.org/legal/resolved.html).
- [ ] (If applicable) I've scanned the updated and new binaries to make sure they do not have vulnerabilities associated with them.

## Pytest Marker Checklist
- [ ] (If applicable) I have added the marker `@pytest.mark.model("<model-type>")` to the new tests which I have added, to specify the Deep Learning model that is used in the test (use `"N/A"` if the test doesn't use a model)
- [ ] (If applicable) I have added the marker `@pytest.mark.integration("<feature-being-tested>")` to the new tests which I have added, to specify the feature that will be tested
- [ ] (If applicable) I have added the marker `@pytest.mark.multinode(<integer-num-nodes>)` to the new tests which I have added, to specify the number of nodes used on a multi-node test
- [ ] (If applicable) I have added the marker `@pytest.mark.processor(<"cpu"/"gpu"/"eia"/"neuron">)` to the new tests which I have added, if a test is specifically applicable to only one processor type

#### EIA/NEURON Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `ei_mode = true` or `neuron_mode = true`

#### Benchmark Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `benchmark_mode = true`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license. I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: pytorch/inference/docker/1.8/py3/Dockerfile.neuron [hunk 1]
```diff
@@ -0,0 +1,158 @@
+FROM ubuntu:18.04
+
+LABEL dlc_major_version="1"
+LABEL maintainer="Amazon AI"
+LABEL com.amazonaws.sagemaker.capabilities.accept-bind-to-port=true
+
+ARG PYTHON=python3
+ARG PYTHON_VERSION=3.6.13
+ARG TS_VERSION=0.3.1
+ARG PT_TORCHVISION_URL=https://torchvision-build.s3-us-west-2.amazonaws.com/1.7.1/cpu/torchvision-0.8.2-cp36-cp36m-linux_x86_64.whl
```
Comment (saimidu) [inline pytorch/inference/docker/1.8/py3/Dockerfile.neuron:10]: This seems to target PT 1.7.1

File: pytorch/inference/docker/1.8/py3/Dockerfile.neuron [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+    "pyYAML>=5.4,<5.5" \
+    cython==0.29.12 \
+    falcon==2.0.0 \
+    gunicorn==20.0.4 \
+    gevent==1.4.0 \
+    requests==2.22.0 \
+    grpcio==1.24.1 \
+    protobuf==3.16.0 \
+    gast==0.2.2 \
+    pillow==8.2.0 \
```
Comment (saimidu) [inline pytorch/inference/docker/1.8/py3/Dockerfile.neuron:117]: Please use `"Pillow>=8.3"`

File: pytorch/inference/docker/1.8/py3/Dockerfile.neuron [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+    "awscli<2"
+
+RUN pip install neuron-cc[tensorflow] --extra-index-url https://pip.repos.neuron.amazonaws.com \
+ && pip install "torch-neuron>=1.7.1,<1.7.2" --extra-index-url https://pip.repos.neuron.amazonaws.com \
```
Comment (saimidu) [inline pytorch/inference/docker/1.8/py3/Dockerfile.neuron:121]: This seems to be torch-neuron 1.7.1

File: pytorch/inference/docker/1.8/py3/Dockerfile.neuron [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+RUN HOME_DIR=/root \
+ && curl -o ${HOME_DIR}/oss_compliance.zip https://aws-dlinfra-utilities.s3.amazonaws.com/oss_compliance.zip \
+ && unzip ${HOME_DIR}/oss_compliance.zip -d ${HOME_DIR}/ \
+ && cp ${HOME_DIR}/oss_compliance/test/testOSSCompliance /usr/local/bin/testOSSCompliance \
+ && chmod +x /usr/local/bin/testOSSCompliance \
+ && chmod +x ${HOME_DIR}/oss_compliance/generate_oss_compliance.sh \
+ && ${HOME_DIR}/oss_compliance/generate_oss_compliance.sh ${HOME_DIR} ${PYTHON} \
+ && rm -rf ${HOME_DIR}/oss_compliance*
+
+RUN curl https://aws-dlc-licenses.s3.amazonaws.com/pytorch-1.7/license.txt -o /license.txt
```
Comment (saimidu) [inline pytorch/inference/docker/1.8/py3/Dockerfile.neuron:152]: Please change this to pytorch-1.8

File: pytorch/inference/docker/1.8/py3/Dockerfile.neuron [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+COPY config.properties /home/model-server
+
+RUN chmod +x /usr/local/bin/dockerd-entrypoint.py
+RUN chmod +x /usr/local/bin/neuron-start.sh
+
+ADD https://raw.githubusercontent.com/aws/deep-learning-containers/master/src/deep_learning_container.py /usr/local/bin/deep_learning_container.py
+
+RUN chmod +x /usr/local/bin/deep_learning_container.py
+
+RUN pip install --no-cache-dir "sagemaker-pytorch-inference>=2"
```
Comment (saimidu) [inline pytorch/inference/docker/1.8/py3/Dockerfile.neuron:140]: This dockerfile installs the sagemaker inference toolkit, but there aren't any PyTorch inference SM tests.

File: pytorch/buildspec-1-8-1-neuron.yml [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+      target: config.properties
+
+images:
+  BuildNeuronPTInferencePy3DockerImage:
+    <<: *INFERENCE_REPOSITORY
+    build: &PYTORCH_CPU_INFERENCE_PY3 false
+    image_size_baseline: 10000
+    device_type: &DEVICE_TYPE neuron
+    python_version: &DOCKER_PYTHON_VERSION py3
+    tag_python_version: &TAG_PYTHON_VERSION py36
```
Comment (saimidu) [inline pytorch/buildspec-1-8-1-neuron.yml:34]: Please update the buildspec here with the current version of the buildspec-neuron.yml file.

File: test/sagemaker_tests/pytorch/inference/conftest.py [hunk 7]
```diff
@@ -250,18 +250,23 @@ def fixture_ecr_image(docker_registry, docker_base_name, tag):
 def skip_by_device_type(request, use_gpu, instance_type, accelerator_type):
     is_gpu = use_gpu or instance_type[3] in ['g', 'p']
     is_eia = accelerator_type is not None
+    is_neuron = instance_type == 'ml.inf1.xlarge'
```
Comment (saimidu) [inline test/sagemaker_tests/pytorch/inference/conftest.py:253]: The instance type won't always be `inf1.xlarge`, right?

Could this instead be `instance_type.startswith("ml.inf")`?

File: test/sagemaker_tests/pytorch/inference/integration/sagemaker/test_neuron_hosting.py [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+
+from ...integration import model_neuron_dir, resnet_neuron_script, resnet_neuron_input, resnet_neuron_image_list
+from ...integration.sagemaker.timeout import timeout_and_delete_endpoint
+from .... import invoke_pytorch_helper_function
+
+
+@pytest.mark.model("resnet")
+@pytest.mark.processor("neuron")
+@pytest.mark.neuron_test
+def test_neuron_hosting(framework_version, ecr_image, instance_type, sagemaker_regions):
```
Comment (saimidu) [inline test/sagemaker_tests/pytorch/inference/integration/sagemaker/test_neuron_hosting.py:33]: Could you confirm that this test doesn't run when the image supplied is a non-neuron image or the instance type is non `inf1.*`?

File: test/sagemaker_tests/pytorch/inference/resources/resnet/model_neuron/code/resnet18.py [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+def model_fn(model_dir):
+    """Loads a model. Provides a default implementation.
+    Users can provide customized model_fn() in script.
+    Args:
+        model_dir: a directory where model is saved.
+    Returns: A PyTorch model.
+    """
+    import os
+    import torch
+    import torch.neuron
```
Comment (saimidu) [inline test/sagemaker_tests/pytorch/inference/resources/resnet/model_neuron/code/resnet18.py:13]: Please move these imports out to the top of the script.

File: test/sagemaker_tests/pytorch/inference/integration/sagemaker/test_neuron_hosting.py (hunk 8 above)
Comment (aws-vrnatham) [reply test/sagemaker_tests/pytorch/inference/integration/sagemaker/test_neuron_hosting.py:33]: Tested them locally and looks like they are skipped

```
(dlcneuron) ubuntu@ip-172-31-34-187:~/dlc/deep-learning-containers/test/sagemaker_tests/pytorch/inference$ python3 -m pytest -v in
tegration/sagemaker/test_neuron_hosting.py --region $REGION --docker-base-name pytorch-inference-cpu --tag 1.9.0-cpu-py38-ubuntu20
.04-2021-09-30-00-35-41-pre-push --framework-version 1.9.0 --processor cpu --py-version 3 --aws-id 647554078242 --instance-type "m
l.c4.8xlarge" --sagemaker-region $REGION
====================================================== test session starts =======================================================
platform linux -- Python 3.6.9, pytest-5.3.5, py-1.10.0, pluggy-0.13.1 -- /home/ubuntu/dlc/deep-learning-containers/dlcneuron/bin/
python3
cachedir: .pytest_cache
rootdir: /home/ubuntu/dlc/deep-learning-containers/test/sagemaker_tests/pytorch/inference
plugins: requests-mock-1.7.0, rerunfailures-9.0, xdist-1.34.0, cov-2.12.1, forked-1.3.0
collected 1 item                                                                                                                 

integration/sagemaker/test_neuron_hosting.py::test_neuron_hosting SKIPPED                                                  [100%]
```

```
(dlcneuron) ubuntu@ip-172-31-34-187:~/dlc/deep-learning-containers/test/sagemaker_tests/pytorch/inference$ python3 -m pytest -v integration/sagemaker/test_mnist.py --region $REGION --docker-base-name pytorch-inference-neuron --tag 1.8.1-neuron-py37-ubuntu18.04-2021-09-29-18-18-42 --framework-version 1.8.1 --processor neuron --py-version 3 --aws-id 647554078242 --instance-type "ml.inf1.xlarge" --sagemaker-region $REGION
====================================================== test session starts =======================================================
platform linux -- Python 3.6.9, pytest-5.3.5, py-1.10.0, pluggy-0.13.1 -- /home/ubuntu/dlc/deep-learning-containers/dlcneuron/bin/python3
cachedir: .pytest_cache
rootdir: /home/ubuntu/dlc/deep-learning-containers/test/sagemaker_tests/pytorch/inference
plugins: requests-mock-1.7.0, rerunfailures-9.0, xdist-1.34.0, cov-2.12.1, forked-1.3.0
collected 3 items                                                                                                                

integration/sagemaker/test_mnist.py::test_mnist_distributed_cpu SKIPPED                                                    [ 33%]
integration/sagemaker/test_mnist.py::test_mnist_distributed_gpu SKIPPED                                                    [ 66%]
integration/sagemaker/test_mnist.py::test_mnist_eia SKIPPED                                                                [100%]
```

File: test/sagemaker_tests/pytorch/inference/conftest.py (hunk 7 above)
Comment (aws-vrnatham) [reply test/sagemaker_tests/pytorch/inference/conftest.py:253]: assign_sagemaker_remote_job_instance_type is assigning only ml.inf1.xlarge for neuron but yes will change as we have the option to provide --instance-type also.

**Review by saimidu — COMMENTED** [review]
Soft approved. Once tests are complete, let's revert config file changes, and then merge.

**Review by saimidu — APPROVED** [review]

**Pull request merged**